Add time info to health data

anjalshireesh opened the pull request:
## Description

Add the current time from all the nodes, along with the roundtrip time required
to fetch it to the health report data, so that this can be used to detect NTP
configuration issues by `health-analyzer.
  
Return the data under `sys -> config -> time-info`

Sample data of the `config` node:

```json
          "config": { 
            "rlimit-max": 1048576,
            "time-info": {
              "current_time": "2022-01-05T16:15:31.445509567Z",
              "roundtrip_duration": 48
            }
          }
```

## Motivation and Context

To detect if NTP is not configured and alert the user about the same.

## How to test this PR?

- Create a minio cluster involving nodes that have different system time (delta of > 10sec)
- Generate health report for this cluster using `mc admin subnet health {alias} --airgap`
- Verify that the health report JSON contains a new node `time-info` under the `sys -> config` section for each of the servers

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

anjalshireesh committed:
Detect and report NTP configuration issues

Get the current time from all the nodes, and check if the difference
between min and max values is more than an acceptable tolerance (10sec).

Return an error under `sys -> errors` in the health report if this is
the case.

harshavardhana reviewed (changes requested):
https://github.com/minio/warp/blob/master/cli/benchserver.go#L348

check how warp handles this @anjalshireesh - I think perhaps doing this is sufficient in this case.

harshavardhana commented on cmd/utils.go:
```diff
@@ -1139,3 +1139,43 @@ func newTLSConfig(getCert certs.GetCertificateFunc) *tls.Config {
+	}
+
+	return madmin.SysErrors{}
+}
```
A logic of something like this might be nice

```go
			roundtrip := time.Since(sent)
			// Add 50% of the roundtrip.
			delta := time.Since(resp.Time.Add(roundtrip / 2))
			if delta < 0 {
				delta = -delta
			}
			if delta > time.Second {
				return fmt.Errorf("host %v time delta too big (%v). Roundtrip took %v. Synchronize clock on client and retry", host, delta.Round(time.Millisecond), roundtrip.Round(time.Millisecond))
			}
			return nil
```

anjalshireesh replied:
As per the latest discussions, error checking will be done by `health-analyzer`. So this code has been removed from this PR.

anjalshireesh committed:
Use 1 minute threshold instead of just 10 seconds

klauspost reviewed (commented):
In overall terms I don't really understand why we do the processing serverside, and not just report the times as-is per server and leave the processing to subnet?

That way we can tweak detection and warnings/errors on the fly. A latency for the calls will allow us to take that into account as well.

anjalshireesh committed:
Add time info to health report

Instead of performing the NTP check in minio itself, just dump the time
info in the health report under `sys -> config` so that
`health-analyzer` can perform the NTP check.

klauspost commented on cmd/peer-rest-client.go:
```diff
@@ -422,12 +422,21 @@ func (client *peerRESTClient) GetSELinuxInfo(ctx context.Context) (info madmin.S
 	if err != nil {
 		return
 	}
+	roundtrip := int32(time.Since(sent).Seconds())
```
Add as milliseconds.

anjalshireesh replied:
Done

klauspost reviewed (changes requested):
Also needs a merge with master.

anjalshireesh committed:
Capture roundtrip duration in milliseconds

klauspost reviewed (approved):
👍🏼
